src/ripple/basics/CompressionAlgorithms.h

```diff
@@ -0,0 +1,233 @@
+    std::array<std::uint8_t, varint_traits<std::uint32_t>::max> vi;
+
+    if (in_size > UINT32_MAX)
+        do_throw("lz4f compress: invalid size");
+
+    auto const original_size_bytes = write_varint(vi.data(), in_size);
+
+    std::size_t const out_capacity = LZ4F_compressFrameBound(in_size, NULL);
+
+    auto* compressed = bf(original_size_bytes + out_capacity);
+    result.first = compressed;
+
+    std::memcpy(compressed, vi.data(), original_size_bytes);
+
+    size_t compressed_size = LZ4F_compressFrame(
+            compressed + original_size_bytes,
+            out_capacity,
+            in,
+            in_size,
+            NULL);
+    if (LZ4F_isError(compressed_size))
+        do_throw("lz4f failed compress update");
+
+    result.second = original_size_bytes + compressed_size;
+
+    return result;
+}
+
+template<typename InputStream>
+bool
+copy_stream(InputStream& in, std::uint8_t* dst, int size, std::vector<int> &used_size)
+{
+    const void * data = NULL;
+    int data_size = 0;
+    int copied = 0;
+
+    while (copied != size && in.Next(&data, &data_size))
+    {
+        auto sz = data_size >= (size - copied) ? (size - copied) : data_size;
+        std::memcpy(dst + copied, data, sz);
+        copied += sz;
+        used_size.push_back(data_size);
+    }
+
+    return copied == size;
+}
+
+template<typename InputStream>
+std::size_t
+get_original_size(InputStream& in)
+{
+    std::array<std::uint8_t, varint_traits<std::uint32_t>::max> vi;
+    std::size_t original_size = 0;
+    const void * data = NULL;
+    int data_size = 0;
+    std::vector<int> used_size;
+
+    if (!in.Next(&data, &data_size))
+        do_throw("lz4f decompress: invalid input size");
+
+    used_size.push_back(data_size);
+
+    const void *p = data;
+
+    if (data_size < varint_traits<std::uint32_t>::max)
+    {
+        std::memcpy(vi.data(), data, data_size);
+        if (!copy_stream(in, vi.data() + data_size,
+                varint_traits<std::uint32_t>::max - data_size, used_size))
+            do_throw("lz4f decompress: header");
+        p = vi.data();
+    }
+
+    auto original_size_bytes = read_varint(p, varint_traits<std::uint32_t>::max, original_size);
+
+    if (original_size_bytes == 0)
+        do_throw("lz4f decompress:: original_size_bytes == 0");
+
+    // rewind the stream
+    for (auto it = used_size.rbegin(); it != used_size.rend(); ++it)
+        in.BackUp(*it);
+    in.Skip(original_size_bytes);
+
+    return original_size;
+}
+
+template<typename InputStream>
+bool
+next_chunk(InputStream &in, std::size_t in_size,
+        void const *& compressed_chunk, int &chunk_size,
+        std::vector<std::uint8_t> &buffer)
+{
+    bool res = in.Next(&compressed_chunk, &chunk_size);
+
+    if (!res && buffer.size() == 0)
+        return false;
+
+    if (buffer.size() != 0)
+    {
+        auto sz = buffer.size();
+        chunk_size = ((chunk_size + sz) <= in_size) ? chunk_size : (in_size - sz);
```

Comment:
We use ZeroCopyInputStream to get the data from the boost's multiple buffers. We are only going to read the payload size of bytes from the stream and return to the caller (PeerImp calls invokeProtocolMessage) the number of bytes that we read. PeerImp then calls `consume` to consume the number of bytes read from the multiple buffer.

Reply:
I guess I'm still confused on this issue. It seems like we're leaving `InputStream` in an inconsistent state in the sense if `Next` is called again and we don't rewind here then we skipped bytes.

I understand the current callers won't run across this issue. They will never call `Next` after this function returns. That might be OK for some private function of limited scope, but since this is a general-purpose decompressor I'd this we should leave the stream in a consistent state.

Reply:
Fixed.